src/Bicep.Decompiler/Rewriters/ParentChildResourceNameRewriter.cs

```diff
@@ -0,0 +1,174 @@
+                return null;
+            }
+
+            for (var i = 0; i < parent.Expressions.Length; i++)
+            {
+                var childSymbol = semanticModel.GetSymbolInfo(child.Expressions[i]);
+                var parentSymbol = semanticModel.GetSymbolInfo(child.Expressions[i]);
+
+                if (childSymbol == null || childSymbol != parentSymbol)
+                {
+                    return null;
+                }
+            }
+
+            for (var i = 0; i < parent.SegmentValues.Length - 1; i++)
+            {
+                if (child.SegmentValues[i] != parent.SegmentValues[i])
+                {
+                    return null;
+                }
+            }
+
+            var finalIndex = parent.SegmentValues.Length - 1;
+            if (!child.SegmentValues[finalIndex].StartsWith(parent.SegmentValues[finalIndex], StringComparison.Ordinal))
+            {
+                return null;
+            }
+
+            var finalSegmentSuffix = child.SegmentValues[finalIndex].Substring(parent.SegmentValues[finalIndex].Length);
+            if (finalSegmentSuffix.Length == 0 || finalSegmentSuffix[0] != '/')
+            {
+                return null;
+            }
+
+            var parentResourceIdentifier = new Token(TokenType.Identifier, new TextSpan(0, 0), parentResourceSymbol.Name, Enumerable.Empty<SyntaxTrivia>(), Enumerable.Empty<SyntaxTrivia>());
+            var nameProperty = new Token(TokenType.Identifier, new TextSpan(0, 0), "name", Enumerable.Empty<SyntaxTrivia>(), Enumerable.Empty<SyntaxTrivia>());
+            var replacementExpression = new PropertyAccessSyntax(
+                new VariableAccessSyntax(new IdentifierSyntax(parentResourceIdentifier)),
+                new Token(TokenType.Dot, new TextSpan(0, 0), ".", Enumerable.Empty<SyntaxTrivia>(), Enumerable.Empty<SyntaxTrivia>()),
+                new IdentifierSyntax(nameProperty));
+
+            var leftStringToken = new Token(TokenType.StringLeftPiece, new TextSpan(0, 0), "'${", Enumerable.Empty<SyntaxTrivia>(), Enumerable.Empty<SyntaxTrivia>());
+            var nextStringToken = parent.SegmentValues.Length == child.SegmentValues.Length ? 
+                new Token(TokenType.StringRightPiece, new TextSpan(0, 0), StringUtils.EscapeBicepString(finalSegmentSuffix, "}", "'"), Enumerable.Empty<SyntaxTrivia>(), Enumerable.Empty<SyntaxTrivia>()) :
+                new Token(TokenType.StringMiddlePiece, new TextSpan(0, 0), StringUtils.EscapeBicepString(finalSegmentSuffix, "}", "${"), Enumerable.Empty<SyntaxTrivia>(), Enumerable.Empty<SyntaxTrivia>());
+            return new StringSyntax(
+                new [] { leftStringToken, nextStringToken }.Concat(child.StringTokens.Skip(parent.StringTokens.Length)),
+                replacementExpression.AsEnumerable().Concat(child.Expressions.Skip(parent.Expressions.Length)),
+                new [] { "", finalSegmentSuffix }.Concat(child.SegmentValues.Skip(parent.SegmentValues.Length)));
```

Comment:
We could really do with a nice library for this sort of syntax-munging. I didn't want to do it in this PR though.